### PR TITLE
Level3 io solution

### PR DESCRIPTION
**Pull request by waldemar99zielinski**
Server:
- private members:
    - _port
    - _socket - udp4 socket
    - _clients - clients dictionary
- methods:
    - public listen() - creates dgram socket, binds it to default port 8000 or other if given as parameter, sets up dgram handlers, executes callback if given
    - shutdown() - closes dgram socket, executes callback if given
    - private _handle_incoming_message() - based on received message type fire a certain action
    - private _client_register() - adds new client to __client_ dictionary 
    - private _client_leave() - removes client from __client_ dictionary
    - private _client_message() - forwards the payload (using __send_message_ method) to destination specified in received message 
    - private _client_broadcast() - forwards the payload (using __send_message_ method) to all the clients in _clients 
    - private _send_message() - checks if destination client is in __clients_ dictionary, creates message and sends it via udp4 socket send method

Client:
- private members:
    -  _id
    - _username
    - _socket - udp4 socket
    - _server - server's address (Address interface)
- methods:
    - constructor() - assign id, username and creates dgram udp4 socket 
    - public connect() - binds server address to __{localhost, 8000}_ by default or other if given, sets dgram handler to listen for messages, sends message (type REGISTRATION) to the server, returns Promise which resolves with server address
    - public disconnect() - creates message (type LEAVE), sends it to the server via __prepare_and_send_message_, close __socket_, instantiate another udp4 socket
    - public send() - prepares message (_IMessage_ interface with MESSAGE type) which contains given string message and destination number, then sends the message via __prepare_and_send_message_
    - public broadcast() -  prepares message (_IMessage_ interface with BROADCAST type) which contains given string message, then sends the message via __prepare_and_send_message_
    - private _prepare_and_send_message() - takes IMessage as parameter, stringify and sends it to the server using socket send method

### CODE REVIEW

**Comment by lurenzsgp** [comment]
Can you please improve the description?
200w

File: src/Client.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    private _server: Address;
+
+    constructor(id: number, username: string)
+    {
+        this._id = id;
+        this._username = username;
+        this._socket = dgram.createSocket('udp4');
+    }
+
+    public connect(server: Address = {ip: 'localhost', port: 8000}): Promise <Address>
```
Comment (lurenzsgp) [inline src/Client.ts]: 🙈 Check the spaces

File: src/Client.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+        const register_message: IMessage = {
+            type: MessageType.REGISTRATION,
+            source: {
+                id: this._id,
+                username: this._username
+            }
+        };
+
+        //listen to messges from server
```
Comment (lurenzsgp) [inline src/Client.ts]: 🙈
```suggestion
        // listen to messges from server
```

File: src/Client.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        this._socket.on('message', (msg, rinfo) => 
+        {
+            const received_message: IMessage = JSON.parse(msg.toString());
+            console.log(`[CLIENT-${this._id}]: received message: ${received_message.payload} from client ${received_message.source.id}`);
+        });
+      
+       return this._prepare_and_send_message(register_message).then(() =>
+       {
+           return this._server;
+       });
```
Comment (lurenzsgp) [inline src/Client.ts]:
```suggestion
       await this._prepare_and_send_message(register_message);
       
       return this._server;
```

File: src/Client.ts (hunk 1 above)
Comment (lurenzsgp) [inline src/Client.ts]: Why are we returning the address that we receive in input?
It is also stored in the class

File: src/Client.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    public send(message: string, to: number): Promise <void>
+    {
+        const send_message: IMessage = {
+            type: MessageType.MESSAGE,
+            source: {
+                id: this._id,
+                username: this._username
+            },
```
Comment (lurenzsgp) [inline src/Client.ts]: This part is always repeated in every message. Why don't you add a private method to get the message source?

File: src/Client.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        return new Promise((resolve, reject) => 
+        {
+            this._prepare_and_send_message(leave_message) 
+            this._socket.close();
+            this._socket = dgram.createSocket('udp4');
+            return resolve();
+        });
+    }
+
+    public send(message: string, to: number): Promise <void>
```
Comment (lurenzsgp) [inline src/Client.ts]:
```suggestion
    public send(message: string, destination: number): Promise <void>
```

File: src/Client.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            },
+            payload: message
+        };
+
+        return this._prepare_and_send_message(broadcast_message);
+    }
+
+    private _prepare_and_send_message(message: IMessage): Promise<void>
+    {
+        const message_string = JSON.stringify(message);
```
Comment (lurenzsgp) [inline src/Client.ts]: It is always a good practice to abstract from the type and focus on what you are really doing
```suggestion
        const encoded_message = JSON.stringify(message);
```

File: src/Server.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    {
+        this._port = port && typeof port === 'number' ? port : this._port_default_number;
+        this._socket = dgram.createSocket('udp4');
+
+        this._socket.on('error', (error) =>
+        {
+            this._socket.close();
+        });
+
+        this._socket.on('message', (msg, rinfo) =>
```
Comment (lurenzsgp) [inline src/Server.ts]: 🙈 Use extended names

File: src/Server.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+import {IMessage, MessageType} from './Interfaces';
+
+export default class Server
+{
+    private _port: number;
+    private readonly _port_default_number = 8000;
+    private _socket: dgram.Socket;
+    private _clients: {[id: number]: IClient} = {};
+
+    public listen(port?: number | ((port: number) => void), callback?: (port: number) => void)
```
Comment (lurenzsgp) [inline src/Server.ts]:
```suggestion
    public listen(config: {port?: numbe, callback?: (port: number) => void})
```

File: src/Server.ts [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    public shutdown(callback?: () => void)
+    {
+        if(!this._socket)
+            return;
+
+        this._socket.close(() =>
+        {
+            this._socket = dgram.createSocket('udp4');
```
Comment (lurenzsgp) [inline src/Server.ts]: Why do we create a socket in the shutdown method?
Are we preparing a new socked to handle future connections?

File: src/Server.ts [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+                this._client_broadcast(message);
+                break;
+            
+            default:
+                break;
+
+        }
+    }
+
+    private _client_register(message: IMessage, rinfo: dgram.RemoteInfo)
```
Comment (lurenzsgp) [inline src/Server.ts]: What does `rinfo` stay for?

File: src/Server.ts [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        const {address, port} = rinfo;
+
+        this._clients[id] = {
+            id: id, 
+            username: username, 
+            address: {
+                ip: address, 
+                port: port
+            }
+        };
```
Comment (lurenzsgp) [inline src/Server.ts]:
```suggestion
        const {address: ip, port} = rinfo;

        this._clients[id] = {id, username, address: {ip, port}};
```

**Review by lurenzsgp — CHANGES_REQUESTED** [review]

File: src/Server.ts (hunk 9 above)
Comment (waldemar99zielinski) [reply src/Server.ts]: `Are we preparing a new socked to handle future connections?` - that was the idea, but it is all done in the listen method, so I will remove it.

File: src/Client.ts (hunk 1 above)
Comment (waldemar99zielinski) [reply src/Client.ts]: Stated in the task:
` It returns a Promise which resolves with the server's address when the connection is established.`

Comment (lurenzsgp) [reply src/Client.ts]: It was not clear from the description

File: src/Client.ts [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    public async connect(server: Address = {ip: 'localhost', port: 8000}): Promise <Address>
+    {
+        this._server = server;
+
+        const register_message: IMessage = {
+            type: MessageType.REGISTRATION,
+            source: {
+                id: this._id,
+                username: this._username
+            }
```
Comment (lurenzsgp) [inline src/Client.ts]:
```suggestion
            source: this._get_message_source(),
```

File: src/Server.ts [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+            console.log(`[SERVER] destination:${destination} not found`);
+            return;
+        }
+        
+        const message: IMessage = {
+            type: MessageType.MESSAGE,
+            source: {
+                    id: source_id,
+                    username: source_username
+                },
```
Comment (lurenzsgp) [inline src/Server.ts]: 🙈  Check the indentation

**Review by lurenzsgp — CHANGES_REQUESTED** [review]

**Review by lurenzsgp — APPROVED** [review]

**Pull request closed without merging**